### PR TITLE
fix(dataframe): handle pandas dimensionality reduction in .xs() for single-item matches

### LINKED ISSUE #28559: [Bug]:  Some methods on DeferredSeries and DeferredDataFrame don't work right when returning single items
### What happened?

I've been upgrading dataframe to be compliant with Pandas 2, and a new doctest in Pandas 2 revealed this bug that currently exists in Pandas 1.

Whenever `.xs()` will yield a result that is a single item, it will either fail or have the wrong outcome compared to pandas xs. For examples, I have a draft PR with failing tests [here](https://github.com/caneff/beam/pull/1)

There are two different types of failures. There are the cases where we drop dimensionality (like returning a scalar for ser.xs() or a series for `df.xs())` when pick exactly one element for a single Index series/dataframe.  Then we get concatenation failures when trying to combine everything back together, because we are expecting to concat series/frame elements (for ser.xs/frame.xs respectively) but instead recieve a scalar/series because pandas has done the dimension reduction.

The other type of failure is with a MultiIndex, where we call  df.xs() with a tuple the same length as the number of dimensions of the MultiIndex, and only a single item is returned. We still get a df but it looks wrong and has the wrong shape.  

Both types of failure examples are in my link above.

For the series case, when it fails the stacktrace is: 

```python
_____________________________________________________________________ DeferredFrameTest.test_series_xs ______________________________________________________________________

self = <apache_beam.dataframe.frames_test.DeferredFrameTest testMethod=test_series_xs>

    def test_series_xs(self):
      # pandas doctests only verify DataFrame.xs, here we verify Series.xs as well
      d = {
          'num_legs': [4, 4, 2, 2],
          'num_wings': [0, 0, 2, 2],
          'class': ['mammal', 'mammal', 'mammal', 'bird'],
          'animal': ['cat', 'dog', 'bat', 'penguin'],
          'locomotion': ['walks', 'walks', 'flies', 'walks']
      }
      df = pd.DataFrame(data=d)
      df = df.set_index(['class', 'animal', 'locomotion'])
    
      self._run_test(lambda df: df.num_legs.xs('mammal'), df)
      self._run_test(lambda df: df.num_legs.xs(('mammal', 'dog')), df)
      self._run_test(lambda df: df.num_legs.xs('cat', level=1), df)
      self._run_test(
          lambda df: df.num_legs.xs(('bird', 'walks'), level=[0, 'locomotion']),
          df)
    
      df_single_index = df.reset_index().set_index('class')
    
      # Passes because Pandas xs returns a series (multiple matches for 'mammal')
>     self._run_test(lambda df: df.num_legs.xs('mammal'), df_single_index)

frames_test.py:325: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
frames_test.py:191: in _run_test
    expected = expected.sort_values(list(expected.columns))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = class
mammal    4
mammal    4
mammal    2
Name: num_legs, dtype: int64, name = 'columns'

    def __getattr__(self, name: str):
        """
        After regular attribute access, try looking up the name
        This allows simpler access to columns for interactive use.
        """
        # Note: obj.x will always call obj.__getattribute__('x') prior to
        # calling obj.__getattr__('x').
        if (
            name not in self._internal_names_set
            and name not in self._metadata
            and name not in self._accessors
            and self._info_axis._can_hold_identifiers_and_holds_name(name)
        ):
            return self[name]
>       return object.__getattribute__(self, name)
E       AttributeError: 'Series' object has no attribute 'columns'

../../../../../../.virtualenvs/env/lib/python3.11/site-packages/pandas/core/generic.py:5902: AttributeError

```

And for the frame case, when it fails (as opposed to just returning the wrong thing) the stack trace is:

```python
____________________________________________________________________ DeferredFrameTest.test_dataframe_xs ____________________________________________________________________

self = <apache_beam.dataframe.frames_test.DeferredFrameTest testMethod=test_dataframe_xs>

    def test_dataframe_xs(self):
      # Test cases reported in BEAM-13421
      df = pd.DataFrame(
          np.array([
              ['state', 'day1', 12],
              ['state', 'day1', 1],
              ['state', 'day2', 14],
              ['county', 'day1', 9],
          ]),
          columns=['provider', 'time', 'value'])
    
      # Passes because Pandas xs returns a frame (multiple matches for 'state')
      self._run_test(lambda df: df.xs('state'), df.set_index(['provider']))
      # Fails because Pandas xs returns a series (single match for 'county')
>     self._run_test(lambda df: df.xs('county'), df.set_index(['provider']))

frames_test.py:343: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
frames_test.py:198: in _run_test
    pd.testing.assert_series_equal(expected, actual)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

left = time     day1
value       9
Name: county, dtype: object, right =           0 time value
time   day1  NaN   NaN
value     9  NaN   NaN
cls = <class 'pandas.core.series.Series'>

    def _check_isinstance(left, right, cls):
        """
        Helper method for our assert_* methods that ensures that
        the two objects being compared have the right type before
        proceeding with the comparison.
    
        Parameters
        ----------
        left : The first object being compared.
        right : The second object being compared.
        cls : The class type to check against.
    
        Raises
        ------
        AssertionError : Either `left` or `right` is not an instance of `cls`.
        """
        cls_name = cls.__name__
    
        if not isinstance(left, cls):
            raise AssertionError(
                f"{cls_name} Expected type {cls}, found {type(left)} instead"
            )
        if not isinstance(right, cls):
>           raise AssertionError(
                f"{cls_name} Expected type {cls}, found {type(right)} instead"
            )
E           AssertionError: Series Expected type <class 'pandas.core.series.Series'>, found <class 'pandas.core.frame.DataFrame'> instead

```
### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [X] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by ManvithPanyam**
### What does this PR do?

Fixes `.xs()` on `DeferredDataFrame`/`DeferredSeries` when a key matches
exactly one row and all index levels are selected.

**The bug:** pandas reduces dimensionality on single-item `.xs()` matches
(`DataFrame` → `Series`, `Series` → scalar), but Beam's implementation
assumed a static output shape across partitions. Non-matching partitions
return an empty container of the original type, so when the matching
partition returned a dimensionality-reduced result, cross-partition
`pd.concat` either raised `TypeError` (scalar concat) or silently produced
a corrupted schema (NaN columns from a shape mismatch).

**The fix:** when `key_size >= nlevels`, matching partitions are routed
through a wrapped singleton stage that mirrors pandas' actual runtime
output, then unwrapped to the real return type — instead of assuming the
proxy shape holds at execution time.

**Known limitation (documented in code):** the *proxy* schema (computed
at graph-construction time from a 0-row template) can't know whether a
key will match 1 row or several at runtime, so it always assumes the
dimensionality-reduced type. If a key has duplicate matches, pandas
returns the non-reduced container instead — this is fundamentally
undecidable at proxy time, same class of limitation as `sort_values()`,
`describe()`, and other data-dependent-shape operations already in this
module. Tests exercising duplicate-match keys use `check_proxy=False`
accordingly, with the reasoning documented inline.

### Fixes
Fixes #28559

### Tests
Added regression coverage in `frames_test.py` for all reported failure
modes: single-level index single match, MultiIndex 0-levels-remaining
single match (both unique and duplicate-key datasets), and Series
single-item `.xs()`. Full `frames_test.py` suite: 452 passed, 19 skipped,
zero regressions.

### CODE REVIEW

**Comment by ManvithPanyam** [comment]
@tvalentyn — opened a fix for this. Root cause: Beam's `.xs()` assumed
static output shape across partitions, but pandas collapses dimensionality
on single-row matches, breaking cross-partition concat.

One thing flagged for review: the proxy can't distinguish single-match
vs. duplicate-match keys at graph-construction time (no row data
available yet), so it always predicts the dimensionality-reduced type.
Documented this as a known limitation in the code — same class of
tradeoff as other data-dependent-shape ops in this module. Open to
feedback if there's a cleaner way to handle it.

File: sdks/python/apache_beam/dataframe/frames.py [hunk 1]
```diff
@@ -1091,27 +1091,93 @@ def xs(self, key, axis, level, **kwargs):
...
+
+      proxy_frame = reindexed._expr.proxy()
+      k_val = key_series.iloc[0]
+      if isinstance(proxy_frame, pd.DataFrame):
+        dummy_index = (
+            pd.MultiIndex.from_tuples([k_val], names=proxy_frame.index.names) if
+            isinstance(k_val, tuple) else pd.Index([k_val],
+                                                   name=proxy_frame.index.name))
+        dummy_data = {
+            col: [proxy_frame[col].dtype.type()]
```
Comment (tvalentyn) [inline sdks/python/apache_beam/dataframe/frames.py]: dtype.type()  will crash on some Pandas types like Categorical Type.

will smth like this work? 

```
        proxy_frame = reindexed._expr.proxy()
        
        dummy_index = (
            pd.MultiIndex.from_tuples([k_val], names=proxy_frame.index.names) if
            isinstance(k_val, tuple) else pd.Index([k_val],
                                                   name=proxy_frame.index.name))
        
        dummy_obj = proxy_frame.reindex(dummy_index)
        
        xs_proxy = dummy_obj.xs(k_val, **kwargs)
        if isinstance(xs_proxy, (pd.DataFrame, pd.Series)):
            xs_proxy = xs_proxy.iloc[:0] 
```

**Comment by tvalentyn** [comment]
thanks for the contribution. are you a Beam Dataframes user?

**Comment by ManvithPanyam** [comment]
Not really an active Beam DataFrames user day-to-day — found this while
working through the dataframe module after my last PR (#39581, the
empty-CSV restriction tracker fix). This was the next issue that stood
out while digging deeper into the module.

Thanks for the catch on `dtype.type()` — confirmed it breaks on
Categorical (`TypeError: type.__new__() takes exactly 3 arguments`) and
tz-aware datetime (`TypeError: function missing required argument
'year'`).

Applied your `reindex()` approach for the DataFrame branch — works
cleanly and is simpler than the manual per-column construction. For the
Series branch (single-match → scalar case), `reindex().iloc[0]` alone
silently changes dtype on plain numeric types (e.g. int64 → float64,
since reindex introduces NaN), so I kept `dtype.type()` as the primary
path there and only fall back to `reindex().iloc[0]` on the `TypeError`
extension-dtype case.

Also had to handle proxy indexes with duplicate labels — `reindex()`
raises on non-unique index, so I drop duplicates on the proxy copy
before reindexing (doesn't affect real execution, this is proxy-only).

Added regression tests for Categorical, tz-aware datetime, and nullable
Int64 columns. Full suite: 452 passed, 19 skipped, zero regressions.

**Comment by tvalentyn** [comment]
thanks, please don't squash reviewed and unreviewed code in the same commits as it hides the diff between versions. would you be able to rebase so we have both commits? thanks!

**Comment by ManvithPanyam** [comment]
Done — split into two commits, rebased/pushed. First is the original as you reviewed it, second is just the reindex()-based fix and everything else from this thread.

File: sdks/python/apache_beam/dataframe/frames.py [hunk 2]
```diff
@@ -1091,27 +1091,88 @@ def xs(self, key, axis, level, **kwargs):
...
 
-    return frame_base.DeferredFrame.wrap(
-        expressions.ComputedExpression(
-            'xs',
-            xs_partitioned,
-            [reindexed._expr, key_expr],
-            requires_partition_by=partitionings.Index(list(range(key_size))),
-            # Drops index levels, so partitioning is not preserved
-            preserves_partition_by=partitionings.Singleton()))
+      proxy_frame = reindexed._expr.proxy()
```
Comment (tvalentyn) [inline sdks/python/apache_beam/dataframe/frames.py]: from AI review:

```
Calling proxy_frame.index.drop_duplicates() returns a deduplicated Index object which is shorter than the proxy_frame itself. Assigning it back to proxy_frame.index will raise a ValueError: Length mismatch from pandas.

You can simply slice proxy_frame.iloc[:0] unconditionally. An empty DataFrame/Series is guaranteed to have a uniquely-valued index, meaning .reindex() will always succeed in creating the required NaN dummy row, and it perfectly preserves the names and dtypes. This also eliminates the need to duplicate the logic for pd.DataFrame and pd.Series.


You can simplify the entire proxy generation block (lines 1136 to 1161 in your diff) to this:


        proxy_frame = proxy_frame.iloc[:0]
        dummy_index = (
            pd.MultiIndex.from_tuples([k_val], names=proxy_frame.index.names) if
            isinstance(k_val, tuple) else pd.Index([k_val], name=proxy_frame.index.name))

        if isinstance(proxy_frame, pd.DataFrame):
          dummy_obj = proxy_frame.reindex(dummy_index)
          xs_proxy = dummy_obj.xs(k_val, **kwargs)
          if isinstance(xs_proxy, (pd.DataFrame, pd.Series)):
            xs_proxy = xs_proxy.iloc[:0]
        else:
          try:
            xs_proxy = proxy_frame.dtype.type()
          except TypeError:
            xs_proxy = proxy_frame.reindex(dummy_index).iloc[0]
```

What are your thoughts on this?

**Comment by ManvithPanyam** [comment]
Verified — the reindex()-based approach itself is correct and I've adopted it, simplifying quite a bit (unified dummy_index construction, removed the duplicate per-branch dedup/copy logic).

One nuance: kept the is_unique/drop_duplicates() dedup check, just hoisted once at the top instead of duplicated per-branch. Traced it — pandas caches an IndexEngine on an index once something inspects it (e.g. .is_unique, .loc, .get_loc), and that cache can survive slicing to iloc[:0], leaving is_unique stale as False even on an empty result. This path is reachable in practice (e.g. the existing test harness calls .xs() on the full arg before slicing to an empty proxy), so removing the check entirely would reintroduce a reindex() failure in that case.

Also found and fixed a real bug while testing this: a non-empty user-supplied proxy with duplicate index labels (via to_dataframe(pcoll, proxy=df_with_dups)) crashed the old code with a length-mismatch error, since drop_duplicates() shrinks the index but not the frame before reassignment. Pre-slicing to iloc[:0] before the dedup check (as in your suggestion) fixes this too. Added a regression test for it.

453 passed, 19 skipped, zero regressions.

Note: 4 CI checks are failing on unrelated Bigtable IO tests — a fresh
google-cloud-bigtable 2.44.0 (published Sept 3, 22:30 UTC) broke
DirectRow._get_mutations() handling. Confirmed master's own scheduled
runs are failing identically before and after this PR's CI run, so
it's unrelated to this change.

**Review by tvalentyn — APPROVED** [review]

**Comment by tvalentyn** [comment]
thanks!
